Intro transfer learning for Unit 2

bellabf opened the pull request:
Hi everyone, 

This is high level intro to transfer learning for Unit 2 CNN. I followed the content suggest in this discussion: https://github.com/johko/computer-vision-course/issues/64.

It was a missing/abandoned content from the first sprint (the priority was notebooks).

There are still one or two notebooks from this part that are on being worked on right now, but are still being reviewed in our branch. 

Thanks!

miniMaddy commented on chapters/en/Unit 2 - Convolutional Neural Networks/intro-transfer-learning.mdx:
It should be Resources

miniMaddy commented on chapters/en/Unit 2 - Convolutional Neural Networks/intro-transfer-learning.mdx:
```diff
@@ -0,0 +1,35 @@
+# Transfer Learning
+Before we can dig into the details of what transfer learning and fine-tuning mean for neural networks, let’s take musical instruments as an example. The theremin is an electronic musical instrument that makes an eerie sound, commonly associated with thrillers and horror movies. It is very hard to play because it requires you to move both your hands in the air between two antennae to control the p ...
```
```suggestion
Below you see from left to right: the theremin, tannerin, and the trombon.
```
Please add a comma after Below

miniMaddy commented on chapters/en/Unit 2 - Convolutional Neural Networks/intro-transfer-learning.mdx:
```suggestion
Yet, the saxophone player is not starting from zero. He is familiar with things like music theory, rhythm, and timing. These general skills give them an edge over someone who never played any instrument at all.  The act of playing an instrument gives all players a general set of skills that are useful across instruments. This generalization across domains (in our example, musical instruments) is what makes the model learn much faster as opposed to training from zero.
```
Adding a comma

miniMaddy commented:
Great content @bellabf , I added a few changes. I would also suggest to maybe add the hugging face models (https://huggingface.co/models) link or some other link for pretrained models.

bellabf commented:
> Great content @bellabf , I added a few changes. I would also suggest to maybe add the hugging face models (https://huggingface.co/models) link or some other link for pretrained models.

Thanks for the quick review and kind words :hugs:! I've addressed all your comments and committed the suggest changes.

ATaylorAerospace commented:
> Hi everyone,
> 
> This is high level intro to transfer learning for Unit 2 CNN. I followed the content suggest in this discussion: #64.
> 
> It was a missing/abandoned content from the first sprint (the priority was notebooks).
> 
> There are still one or two notebooks from this part that are on being worked on right now, but are still being reviewed in our branch.
> 
> Thanks!

Hello @bellabf ! Do you want us to review/approve now or wait until the realted notebooks are completed?

merveenoyan commented on chapters/en/Unit 2 - Convolutional Neural Networks/intro-transfer-learning.mdx:
```suggestion
## Transfer Learning and Fine-tuning
```

merveenoyan commented on chapters/en/Unit 2 - Convolutional Neural Networks/intro-transfer-learning.mdx:
```suggestion
Let's make a distinction between the concepts we are talking about. The trombone player needs no training to play the tannerin. He already knows how to do it unbeknownst to him. The saxophone player needs some training to fine-tune his skills to play the tannerin. In deep learning terms, the trombone player uses a model off-the-shelf. This is called transfer learning. The training of a model that needs more time to learn, like our saxophone player, is called fine-tuning. 
```

merveenoyan reviewed (commented):
Markdown part is great! just left wording nits

merveenoyan commented on notebooks/Unit 2 - Convolutional Neural Networks/timm_Resnet.ipynb:
```diff
@@ -0,0 +1,1187 @@
+{
```
<p>we could say "We will now initialize the cross entropy loss function and the SGD optimizer."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/198/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

merveenoyan replied:
<p>Fine-tuning the Model*</p><p>Also would be nice to have few explanations below the header</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/198/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

merveenoyan replied:
<p>Same here, would be nice to have few explanations</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/198/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

merveenoyan commented:
@bellabf thanks a lot for picking up this section 🦸‍♀️

bellabf commented:
> > Hi everyone,
> > This is high level intro to transfer learning for Unit 2 CNN. I followed the content suggest in this discussion: #64.
> > It was a missing/abandoned content from the first sprint (the priority was notebooks).
> > There are still one or two notebooks from this part that are on being worked on right now, but are still being reviewed in our branch.
> > Thanks!
> 
> Hello @bellabf ! Do you want us to review/approve now or wait until the related notebooks are completed?

Sorry, I missed this comment early. I've sent the .mdx early to review it faster. I've tagged you because you were the person who had made a lot of suggestions in the initial outline discussion. We will focus on reviewing the notebook now! Thanks :smile:

ATaylorAerospace reviewed (commented):
Hello @bellabf I did not see any additional nits outside of items that @merveenoyan allready noted :-)

johko commented on chapters/en/Unit 2 - Convolutional Neural Networks/intro-transfer-learning.mdx:
```suggestion
Transfer learning requires that the previous knowledge is “useful” for the other task. Thus, the features we are trying to explore need to be general enough for the new application. If we go back to our musical instrument example, playing the saxophone instead of the trombone is not as helpful in learning how to play the tannerin. The main skill that gives the trombone player its head start is the intuitive understanding of where the slide should be. 
```

johko commented on chapters/en/Unit 2 - Convolutional Neural Networks/intro-transfer-learning.mdx:
```suggestion
Our example also gives us an interesting nuance. The theremin was too hard to play, so they invented an easier instrument that produced the same sound. The output is nearly the same but needs a lot less training time. For computer vision, we might first do object detection to see where a dog is within an image, and then build a classifier to tell us which breed of dog instead of trying to build a classifier right away.
```

johko commented on chapters/en/Unit 2 - Convolutional Neural Networks/intro-transfer-learning.mdx:
```suggestion
Finally, transfer learning is not a universal performance enhancer. In our example, playing an instrument might help us learn another one, but it might also hinder progress. There are patterns and vices from one instrument that might slow down the progress of another one. If these vices are deeply entrenched within the player, a novice player might surpass the new player with the same amount of training. If your players are stuck to their vices, it might be time to hire new ones.
```

johko commented on chapters/en/Unit 2 - Convolutional Neural Networks/intro-transfer-learning.mdx:
```suggestion
Transfer learning shines especially when there is not enough labeled data to retrain a model from scratch. Using our example, we can think that given enough time, a player who attends just a few lessons can learn on their own by playing the instrument without the constant supervision of their professor. Learning, partially or entirely, on your own in deep learning is called self-training. It allows us to train the model using both labeled (the lessons) and unlabeled (the players on their own) data to learn the task.  
```

johko reviewed (commented):
Nice work on the mdx part. Really well explained with the instrument parallels

ratan commented:
Nice work @bellabf.
All the minor nits are already mentioned by @merveenoyan @johko. Looks good

ratan reviewed (commented):
Looks good

michaelshekasta reviewed (approved):
I think that the explanation is very clear
Pay attention - I don't check the notebooks

merveenoyan commented:
@bellabf can you run the styling scripts and we can merge?

aman06012003 commented:
I have updated the notebook as per the sugesstions from @merveenoyan

bellabf commented:
> @bellabf can you run the styling scripts and we can merge?

Hi, Merve! I am not sure what/where is the styling scripts. Can you point them out to me? Thank you so much.

merveenoyan commented:
@bellabf apparently style checks are passing, you just need to add this to toctree and we'll merge 😊

merveenoyan commented:
@bellabf if you can implement the following changes, the build one will go green 
https://github.com/johko/computer-vision-course/pull/223/commits/8b410ff9b0210a0dbe86f540a040141f103e22c3
https://github.com/johko/computer-vision-course/pull/223/commits/8b410ff9b0210a0dbe86f540a040141f103e22c3
https://github.com/johko/computer-vision-course/pull/223/commits/bd5673b36b8b75a39e941a211ba10c023c349399
https://github.com/johko/computer-vision-course/pull/223/commits/63cabaa80138bc3128b00163980576d58a19ce8f

bellabf commented:
@merveenoyan Hi, Merve. I was away for these last few days, thank you for taking care of this :hugs:.
